### PR TITLE
Add documentation of OpenTelemetry setup with the collector

### PR DESCRIPTION
**Pull request by keiko713**
https://pganalyze-we-docs-updat-lyhnyi.herokuapp.com/docs/collector/otel-setup

Add some starter documentation for OpenTelemetry setup with the collector. Following is the list of what currently covered and ideas of what we could cover in the future. I _think_ what we have is okay enough as a start but happy to hear otherwise.

Third party
* ✅ Honeycomb
* ✅ New Relic
* Jaeger
* Sentry

Language
* ✅ Ruby
* ✅ Go
* ✅ Python
* Node.js
* Java

### CODE REVIEW

**Comment by keiko713** [comment]
@lfittl This is still WIP, but I wanted you to check quickly (toc level is fine) to see if this aligns with what you're thinking. Thanks!

**Comment by keiko713** [comment]
Adding @msakrejda as a reviewer too in case he can look at this with "a fresh eye who is not familiar with OpenTelemetry", esp. overview part.

File: collector/otel-setup/newrelic.mdx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+### Start sending data of your application to New Relic
+
+To start sending data of your application to New Relic, you can either use the
+guided install and install the New Relic agent, or use OpenTelemetry SDKs.
+You can read how to setup OpenTelemetry integration in [their documentation](https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/get-started/opentelemetry-set-up-your-app/).
+
+Using New Relic agents If you choose to use the New Relic agent, make sure to use the version that
+supports [Distributed Tracing](https://docs.newrelic.com/docs/distributed-tracing/concepts/introduction-distributed-tracing/).
+Latest version of agents in each language support it and also it's enabled by
```
Comment (msakrejda) [inline collector/otel-setup/newrelic.mdx:32]:
```suggestion
The latest versions of agents in each language support it and it's enabled by
```
Since there's multiple agents, there's multiple versions.

File: collector/otel-setup/newrelic.mdx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+### Add traceparent query tag to the query
+
+When you use OpenTelemetry SDKs to send data to New Relic, you can follow the
+[traceparent documentation](/docs/collector/otel-setup/traceparent) to add a
+traceparent query tag to the query.
+
+When you are using the New Relic agent, you'll need to modify how to generate a
+traceparent value. You can still follow the traceparent documentation for how to
+add it as a query tag, then you can generate a traceparent tag like the
+following ways.
```
Comment (msakrejda) [inline collector/otel-setup/newrelic.mdx:44]:
```suggestion
add it as a query tag, then you can generate a traceparent tag as follows.
```

File: collector/otel-setup/newrelic.mdx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+for key, value in dummy_headers:
+    if key == 'traceparent':
+        traceparent = value
+        break
+
+# only add traceparent as a query tag when it's not None
+```
+
+For Go, it is the best to use the OpenTelemetry SDK as it will integrate well
```
Comment (msakrejda) [inline collector/otel-setup/newrelic.mdx:79]:
```suggestion
For Go, it is best to use the OpenTelemetry SDK as it will integrate well
```

File: collector/otel-setup/newrelic.mdx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+with sqlcommenter and the traceparent query tag will be added without any extra
+setups.
```
Comment (msakrejda) [inline collector/otel-setup/newrelic.mdx:81]:
```suggestion
setup.
```

File: collector/otel-setup/traceparent.mdx [hunk 5]
```diff
@@ -0,0 +1,124 @@
+---
+title: 'Setup traceparent query tag within your application'
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:2]:
```suggestion
title: 'Set up traceparent query tag within your application'
```

File: collector/otel-setup/traceparent.mdx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+backlink_href: /docs/collector/otel-setup
+backlink_title: 'OpenTelemetry Setup'
+---
+
+In order for a tracing span created by the pganalyze collector to correctly
+attach to the existing tracing on the application side, the query needs to have
+a `traceparent` query tag, as defined in the [W3 Trace Context specification](https://www.w3.org/TR/trace-context/#traceparent-header).
+
+Query tags are key-value pairs appended to the SQL statements as query comments.
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:11]:
```suggestion
Query tags are key-value pairs appended to SQL statements as query comments.
```

File: collector/otel-setup/traceparent.mdx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+* [sqlcommenter](https://google.github.io/sqlcommenter/) format: `key='value'`
+* sqlcommenter format without single quotes: `key=value`
+* [marginalia](https://github.com/basecamp/marginalia) format: `key:value`
+
+## Add traceparent with Go
+
+The easiest way to add a query tag with Go is to use a drop-in replacement for
+go's `database/sql` library, provided by [sqlcommenter](https://google.github.io/sqlcommenter/go/database_sql/).
+
+That library provides an option `EnableTraceparent`, you can enable it to append
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:24]:
```suggestion
That library provides an option `EnableTraceparent`: you can enable it to append
```

File: collector/otel-setup/traceparent.mdx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+a traceparent query tag as a query comment. Please note that this is only
+available when you're using OpenTelemetry for your application's tracing too, as
+it is going to refer the context and the trace context should be filled by the
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:27]:
```suggestion
it is going to reference the context and the trace context should be filled by the
```

File: collector/otel-setup/traceparent.mdx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        trace_id: span_context.hex_trace_id,
+        span_id: span_context.hex_span_id,
+        trace_flags: trace_flag
+      )
+    }
+  }
+]
+```
+
+With above configuration, the query will have the comment like the following:
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:78]:
```suggestion
With above configuration, queries will have the comments like the following:
```

File: collector/otel-setup/traceparent.mdx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+```sql
+SELECT "articles".* FROM "articles" /*application='myblog',action='index',controller='articles',traceparent='00-953c157ad37a81d94fe83bb95cfc8add-3cf784b99bd43910-01'*/
+```
+
+### Marginalia
+
+If you are using Rails 6 and below, you can attach SQL comments using a
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:86]:
```suggestion
If you are using Rails 6 and below, you can attach SQL comments using the
```

File: collector/otel-setup/traceparent.mdx [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    end
+  end
+end
+
+Marginalia::Comment.components = [:application, :controller, :action, :job, :traceparent]
+```
+
+### Limitations
+
+If you're using the OpenTelemetry SDK, this is best to be used with `OpenTelemetry::Instrumentation::ActiveRecord`
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:116]:
```suggestion
If you're using the OpenTelemetry SDK, this is best used with `OpenTelemetry::Instrumentation::ActiveRecord`
```

File: collector/otel-setup/traceparent.mdx [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+and `OpenTelemetry::Instrumentation::PG` instrumentation. Currently, the query
+comment is attached by either Rails or marginalia after the ActiveRecord
+instrumentation, but before the pg instrumentation.
+This makes the span created by the pganalyze collector to have a span from the
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:120]:
```suggestion
This makes the span created by the pganalyze collector have a span from the
```

File: collector/otel-setup/traceparent.mdx [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+ActiveRecord instrumentation as a parent, not the pg one.
+This will cause the database/query span (from pg) and EXPLAIN Plan (from the
+pganalyze collector) to be a "parallel" relation, which may make it harder to
+identify which database span is associated with which EXPLAIN Plan.
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:124]: I don't fully understand this: if I'm using the SDK, does using the above `OpenTelemetry::Instrumentation` classes avoid this problem, or is this a separate issue?

**Review by msakrejda — COMMENTED** [review]
Some wording comments, but as someone totally unfamiliar with OpenTelemetry, it looks pretty clear (though to be fair, I have not tried the instructions).

I do think we should be clearer in some places, especially in terms of sending tracing data to OT endpoints (I don't think anyone will actually think you need to send application data there, but it doesn't hurt to be explicit) and the OpenTelemetry collector versus our collector.

File: collector/otel-setup.mdx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+title: 'Set up OpenTelemetry exporter with the pganalyze Collector'
+backlink_href: /docs/collector
+backlink_title: 'pganalyze Collector'
+---
+
+To integrate with application performance monitoring (APM) tools, the collector
+can optionally act as an [OpenTelemetry tracing](https://opentelemetry.io/docs/concepts/signals/traces/)
+exporter.
+This adds data from pganalyze into existing application traces, and allows
+application engineers to have a quick way to find out why SQL queries are slow.
```
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup.mdx:11]: spot on, new wordings are definitely better 👍 we're using this wording in settings.mdx too so I'll update there too

File: collector/otel-setup.mdx [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+The exporter will send a tracing span for each [collected EXPLAIN plan](/docs/explain/setup)
+to the configured OpenTelemetry endpoint, with a link back to the full plan in
+pganalyze via the `db.postgresql.plan` span attribute.
+The span will be recorded with a service name "Postgres (pganalyze)" and a span
+name "EXPLAIN Plan" by default.
+
+![Overview of OpenTelemetry setup with the pganalyze collector](otel-overview.svg)
+
+## Setup the pganalyze collector with third-party services
```
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup.mdx:21]: looks like... if it's "XXX setup", I can use a one word "setup", and if it's more about "set up X" (verb), I should use "set up"?

File: collector/otel-setup.mdx [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+By setting up `otel_exporter_otlp_endpoint` (or `OTEL_EXPORTER_OTLP_ENDPOINT`),
+the collector will start sending out tracing spans to the specified endpoint.
+
+You can test this by pointing this endpoint to an OpenTelemetry collector
+instance running locally too. For example, you can use the [collector Docker image](https://hub.docker.com/r/otel/opentelemetry-collector-contrib)
```
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup.mdx:27]: Yeah good point. I think I only mention the OpenTelemetry collector here, so I'll make it extra clear about what I'm talking about.

[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup.mdx:27]: Actually I was thinking about making this as "Setting up with the OpenTelemetry collector" or "Setting up locally".
Maybe having a separate page will help clarifying (so that only that page will talk about the OpenTelemetry collector). I'll think a bit more.

File: collector/otel-setup/honeycomb.mdx [hunk 17]
```diff
@@ -0,0 +1,41 @@
+---
+title: 'Set up the pganalyze collector with Honeycomb'
+backlink_href: /docs/collector/otel-setup
+backlink_title: 'OpenTelemetry Setup'
+---
+
+You can set up the pganalyze collector OpenTelemetry exporter with Honeycomb so
+that you can have a span for the EXPLAIN plan inside of the trace of your
+application, such as a web request trace.
```
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup/honeycomb.mdx:9]: My intention is that it still has some one line overview of what this page is talking about (or a benefit of you completing this integration), in case this page is directly linked (e.g. for honeycomb users).
I'll change the first sentence to your suggestion and will keep the second sentence.

File: collector/otel-setup/honeycomb.mdx [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+1. Start sending data of your application to Honeycomb
+2. Add traceparent query tag to the query (see [Setup traceparent query tag within your application](/docs/collector/otel-setup/traceparent) section for the detail)
+3. Create a Honeycomb API key
+4. Setup a Honeycomb API key and an endpoint in the pganalyze collector
+
+If you have already started sending data of your application to Honeycomb and
+you are fine using the existing API key for the pganalyze collector, you can
+skip step 3 and use that key.
+If you wish to create a new API key instead of using the existing API key of the
+environment, you can create it with a ["Send Events" permission](https://docs.honeycomb.io/working-with-your-data/settings/api-keys/#api-key-permissions).
```
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup/honeycomb.mdx:28]: Interesting, I wasn't thinking "can" imply that there. I meant "if you want, you can create it, but you'll need this permission". Your rewording clarifies this so I'll update 👍

File: collector/otel-setup/traceparent.mdx (hunk 13 above)
Comment (keiko713) [reply collector/otel-setup/traceparent.mdx:124]: Ah good point. It's a separate issue, but it's the issue happens only with these instrumentation classes. I'm going to rewrite this part a bit.

File: collector/otel-setup.mdx (hunk 16 above)
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup.mdx]: Ok I didn't make a new page but I put this into the separate section, "### Test setting up locally with the OpenTelemetry collector". I think it's better as a structure, though I'm not super sure if this is even needed.
This is really about test/dev env and feels like this much testing is not really required by customers (maybe only us).
I'm fine dropping the whole section too if it can be too much (or too confusing). Thoughts? @lfittl

**Comment by keiko713** [comment]
Thanks for the thorough review as usual, Maciek!
Ok I applied review feedback, two major parts are tweaked (re: OpenTelemetry collector and Ruby traceparent limitations). I left the comments unresolved for them in case Maciek wants to take another look, though I think we could also leave that to Lukas too.

File: collector/toc.yml [hunk 19]
```diff
@@ -8,3 +8,15 @@
...
+- mount: collector/otel-setup
+  items:
+  - name: Overview
+    href: collector/otel-setup
+  - name: Set up traceparent query tag
+    href: collector/otel-setup/traceparent
+  - name: OpenTelemetry Setup with Honeycomb
+    href: collector/otel-setup/honeycomb
+  - name: OpenTelemetry Setup with New Relic
+    href: collector/otel-setup/newrelic
```
Comment (lfittl) [inline collector/toc.yml:22]: I was thinking we actually put this at the top-level, to highlight the fact that we integrate with APM providers more explicitly (its a bit hard to find in the collector sub-navigation).

Like this, also emphasizing integrating for the New Relic / Honeycomb pages:

Screenshot 2023-10-24 at 11 03 29 PM

**Review by lfittl — APPROVED** [review]

File: collector/otel-setup/newrelic.mdx [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+1. [Start sending tracing data from your application to New Relic](#start-sending-data-of-your-application-to-new-relic)
+2. [Add traceparent query tag to the query](#add-traceparent-query-tag-to-the-query)
+3. [Create a New Relic API key](#create-a-new-relic-api-key)
+4. [Set up an API key and an endpoint in the pganalyze collector](#setup-an-api-key-and-an-endpoint-in-the-pganalyze-collector)
+
+### Start sending tracing data from your application to New Relic
+
+To start sending tracing data from your application to New Relic, you can either
+use the guided install and install the New Relic agent, or use OpenTelemetry
+SDKs. You can read how to setup OpenTelemetry integration in
```
Comment (msakrejda) [inline collector/otel-setup/newrelic.mdx:27]:
```suggestion
SDKs. You can read how to set up the OpenTelemetry integration in
```
:smile:

File: collector/otel-setup/newrelic.mdx [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+For Go, it is best to use the OpenTelemetry SDK as it will integrate well with
+sqlcommenter and the traceparent query tag will be added without any extra
+setup.
+
+### Create a New Relic API key
+
+You can create a new API key with a ["License" key type](https://docs.newrelic.com/docs/apis/intro-apis/new-relic-api-keys/).
+You can optionally use the same key as the one used by your application.
+
+### Setup an API key and an endpoint in the pganalyze collector
```
Comment (msakrejda) [inline collector/otel-setup/newrelic.mdx:87]:
```suggestion
### Set up an API key and an endpoint in the pganalyze collector
```

File: collector/otel-setup/traceparent.mdx [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+a traceparent query tag as a query comment. Please note that this is only
+available when you're using OpenTelemetry for your application's tracing too, as
+it is going to reference the context and the trace context should be filled by
+the application side already.
+
+## Add traceparent with Ruby
+
+### Prerequisite
+
+Setup OpenTelemetry SDK with your application. Read the
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:34]:
```suggestion
Set up OpenTelemetry SDK with your application. Read the
```

File: collector/otel-setup.mdx (hunk 15 above)
[the remark this answers is not in the document]
Comment (msakrejda) [reply collector/otel-setup.mdx]: Yeah, exactly. "Setup" is a noun, "set up" is the verb. Like "login" and "log in". It's pretty common to use the noun form as a verb, but I haven't seen it in a dictionary yet, so we should probably stick to the standard usage.

File: collector/otel-setup.mdx (hunk 16 above)
[the remark this answers is not in the document]
Comment (msakrejda) [reply collector/otel-setup.mdx]: Yeah, this section is now very clear. Thanks for rewriting.

File: collector/otel-setup/traceparent.mdx [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+integration is best used with `OpenTelemetry::Instrumentation::ActiveRecord` and
+`OpenTelemetry::Instrumentation::PG` instrumentation.
+With this setup, the EXPLAIN Plan span from the pganalyze collector will be
+created as a child trace of the ActiveRecord span, making it easy to relate the
+EXPLAIN Plan span to the ActiveRecord code.
+However, the span from the pg instrumentation will be created as a sibling span
+of the pganalyze collector, which may make it harder to relate them together.
+This is due to the limitation of when we can generate a query comment, which is
+currently only possible after the ActiveRecord instrumentation, but before the
+pg instrumentation.
```
Comment (msakrejda) [inline collector/otel-setup/traceparent.mdx:126]: I still don't quite understand what this is telling me (this is probably due to my lack of familiarity with otel at this point). Specifically, what does it mean that it's created as a sibling span, and can I do anything about that? Also, is this a consequence of using the `OpenTelemetry::Instrumentation` classes, or a separate problem. If the latter should this be two separate limitations bullet points?

No need to edit this further on my account at this point (given I have no relevant otel experience), but I wanted to note some parts that are still unclear to me.

**Review by msakrejda — COMMENTED** [review]
Noted a couple of other "setup" / "set up" issues, but otherwise it looks good. The otel collector section is great.

File: collector/otel-setup/traceparent.mdx (hunk 23 above)
Comment (keiko713) [reply collector/otel-setup/traceparent.mdx:126]: FYI: I chatted with Maciek and explained this. We agreed that this is not needed as it's kinda small detail, so I'll drop this for now.

File: collector/otel-setup/traceparent.mdx [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+a `traceparent` tag:
+
+* [sqlcommenter](https://google.github.io/sqlcommenter/) format: `key='value'`
+* sqlcommenter format without single quotes: `key=value`
+* [marginalia](https://github.com/basecamp/marginalia) format: `key:value`
+
+## Add traceparent with Go
+
+The easiest way to add a query tag with Go is to use a drop-in replacement for
+go's `database/sql` library, provided by [sqlcommenter](https://google.github.io/sqlcommenter/go/database_sql/).
```
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup/traceparent.mdx]: Good point. I aligned this with Ruby side a bit more, and made `database/sql` as a sub section of it (like how I have Rails and Marginalia in Ruby part)

File: collector/otel-setup/newrelic.mdx [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+dummy_headers = []
+newrelic.agent.insert_distributed_trace_headers(dummy_headers)
+traceparent = None
+
+for key, value in dummy_headers:
+    if key == 'traceparent':
+        traceparent = value
+        break
+
+# only add traceparent as a query tag when it's not None
```
[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup/newrelic.mdx]: I've been looking into this, and trying to find out how to do it but actually it doesn't look like super straight forward (docs are not so great, sadly).

Looks like Python provides some auto instruments which auto-detect what packages the app has and enable proper instrumentations. I managed to set up with Flask + Postgres, though to enable, I had to do the following and I'm not so sure if that's the "correct way" to do it:

```
 Psycopg2Instrumentor().uninstrument()
 Psycopg2Instrumentor().instrument(enable_commenter=True, commenter_options={})
 ```

The good news is, OpenTelemetry Python supports emitting the sql comment pretty well like above, you can pass the arg to the instrumentor and the comment will be there. Sad news is, this forces users to use the manual instrument I think. I _think_ I was able to use the hybrid but my flask app is just so tiny and not sure if the hybrid (of auto instruments + some manual) is possible. I also found this GH issue which will make our life much easier, as we only need to ask customers to turn on this ENV var: https://github.com/open-telemetry/opentelemetry-specification/issues/3560

I'm also having problem setting up this with Django :/ I made a super vanilla app and yet it's not working (getting the same error as this person: https://stackoverflow.com/questions/77346904/django-python-you-must-set-settings-allowed-hosts-if-debug-is-false).

tl;dr -- I think I can write something based on above. At least mentioning something like "if you enable comments with instrumentor (you can also specify at framework level too), traceparent will be there automagically", without going into too much detail.
Thoughts?

File: collector/toc.yml (hunk 19 above)
Comment (keiko713) [reply collector/toc.yml]: I've updated the structure, can you please double check and see if this is what you're looking for?

https://pganalyze-we-docs-updat-lyhnyi.herokuapp.com/docs/opentelemetry

Comment (lfittl) [reply collector/toc.yml]: Yes, that is exactly what I was thinking 👍

File: collector/otel-setup/newrelic.mdx (hunk 25 above)
[the remark this answers is not in the document]
Comment (lfittl) [reply collector/otel-setup/newrelic.mdx]: > At least mentioning something like "if you enable comments with instrumentor (you can also specify at framework level too), traceparent will be there automagically", without going into too much detail.

Yeah, I think that's a good starting point - we can revise it based on customer feedback as needed.

[the remark this answers is not in the document]
Comment (keiko713) [reply collector/otel-setup/newrelic.mdx]: Ok I updated. I think I started to understand what's going on but I'm not quite sure.
I wrote "you need to use manual instrumentation" but I think it's not entirely true and you could still do the automatic instruments, then just do uninstrument/instrument the specific instrumentor like I demonstrated above. That's said, I'm really not confident on that so I'll go with "you need to use manual instrumentation" for now. Hope the reader who already has OpenTelemetry instrumentation enabled on their application... knows more things than me 🙈

**Comment by keiko713** [comment]
I'm gonna go ahead and will merge this. If you spotted any typos or can-be-improved part, happy to fix as a follow up.
